rq/worker.py

```diff
@@ -667,24 +669,36 @@ def perform_job(self, job, queue):
                 # to use the same exc handling when pickling fails
                 job._result = rv
 
-                self.set_current_job_id(None, pipeline=pipeline)
-
                 result_ttl = job.get_result_ttl(self.default_result_ttl)
                 if result_ttl != 0:
                     job.ended_at = utcnow()
-                    job.set_status(JobStatus.FINISHED, pipeline=pipeline)
-                    job.save(pipeline=pipeline)
 
-                    finished_job_registry = FinishedJobRegistry(job.origin,
-                                                                self.connection)
-                    finished_job_registry.add(job, result_ttl, pipeline)
+                while True:
+                    try:
+                        # if dependencies are inserted after enqueue_dependents
+                        # a WatchError is thrown by execute()
+                        pipeline.watch(job.dependents_key)
```

Comment:
That does not work: The `try/except` in `enqueue_dependents` was only some kind of "feature". For the usual case, where `enqueue_dependents` is called by the worker, the outer try/except is important, because the `WatchError` is actually raised at L698 (worker.py, below). If only `enqueue_dependents` is watched, there is again a possible race-condition if a job was enqueued afterwards, just before the job is finished ;)
But maybe we could find a compromise: extract the finalizing logic (into `handle_successful_job_execution`) as you suggested, but also move the code of `enqueue_dependents` into it. Why the latter one? The `watch/multi` commands are tightly coupled, but atm `watch` is called in the worker, and `multi` is called in `enqueue_dependents`. Maybe it is just not possible or easy to understand, if this happens in two different places. Merging the code into one new method in the worker could make the code much clearer.